src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -0,0 +1,149 @@
+        }
+
+        protected internal override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventType, PinChangeEventHandler callback)
+        {
+            throw new NotImplementedException();
+        }
+        
+        protected internal override void ClosePin(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pinHandle))
+            {
+                Interop.ReleaseGpiodLine(pinHandle);
+                _pinNumberToSafeLineHandle.Remove(pinNumber);
+            }
+        }
+
+        protected internal override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber) => throw new
+            PlatformNotSupportedException("This driver is generic so it cannot perform conversions between pin numbering schemes.");
+
+        protected internal override PinMode GetPinMode(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pinHandle))
+            {
+                return (Interop.GetLineDirection(pinHandle) == 1) ? PinMode.Input : PinMode.Output;
+            }
+            throw new IOException($"Error while reading pin mode for pin: {pinNumber}");
+        }
+
+        protected internal override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            // Libgpiod Api do not support pull up or pull down resistors for now.
+            return mode != PinMode.InputPullDown && mode != PinMode.InputPullUp;
+        }
+
+        protected internal override void OpenPin(int pinNumber)
+        {
+            SafeLineHandle pinHandle = Interop.GetChipLineByOffset(_chip, pinNumber);
+            if (pinHandle == null)
+            {
+                throw new IOException($"Pin number {pinNumber} not available for chip: {_chip} error: {Marshal.GetLastWin32Error()}");
+            }
+            _pinNumberToSafeLineHandle.Add(pinNumber, pinHandle);
+        }
+
+        protected internal override PinValue Read(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pinHandle))
+            {
+                int value = Interop.GetGpiodLineValue(pinHandle);
+                if (value == -1)
+                {
+                    throw new IOException($"Error while reading value from pin number: {pinNumber}, error: {Marshal.GetLastWin32Error()}");
+                }
+                return value;
+            }
+            throw new IOException($"Error while reading value from pin number: {pinNumber}");
+        }
+
+        protected internal override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            throw new NotImplementedException();
+        }
+
+        protected internal override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            int success = -1;
```

Comment:
nit: rename to result?

Reply:
or make it bool and later `success = Interop....(...) != -1`

Reply:
@ahsonkhan this variable is for determining if setPinMode call was successful, i think if it was returning result we better rename it into result

@krwq that can be improvment, as main goal is if it was failed throw exception maybe change it bool failed

Reply:
> this variable is for determining if setPinMode call was successful, i think if it was returning result we better rename it into result

I was thinking more towards what is the result of the request (i.e. Interop.RequestLineInput) rather than it being successful or not. Naming it as success or failure implies one or the other when it could be either.

Changing it to a bool unnecessarily introduces more branches in the code. and increases the method body size. **We should consider reverting it back to int, imo.** The rename suggestion was a nit so either one of these is fine: result/failed/success/requestResult/etc.

Looking at a subset of the method code:
[51 instructions (using integer) vs 57 (now with bool) vs 42 (best - cache local + use integer) - sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgBgARYCMA3ALABQuBhAdAEoCuAdjAJZhw0DCA9mAA6sANogDKiAG6sAxnChlKAZgIAmPFzwBvCnl0FlWACx5xMAAqsmAWR4ATOAH1+COBIAUlmHkFMAcgzBgRCQ8YB4eITwwOzgASi0dPSTPPCgGaVkoKDwAXjwAWhJEpN1WADM8N2j7WOKS7XISptT0zOy8ujgARwY5GAAZSzg3H39A4O9LMaCEGgAVHlEYBEsAczdY2IVmvABfOqS4ISg4A70GnaS0jLl2vE6evsGmYdGAmZC38dmFpZWmdabbbNfaNZpnUoVNzXNq5PKFWpgpoXS66GAACwQPAA7ngXriAKIAD1k/DYPCYG2BTVBSVpejqWAMxlMFmsMQc+I8LEmfneEzCESiMXiKKagsiZQgwjgtlyeClxzg1JK5Uq1TiEISSMuUplco63V6sGerym/OQvOmiHmi2Waw28Ry8KKOrpWqOJy1Yt10pEBvuRqeQxG5u+nzDM1tfwdmzhBVdO3pTS1arcev9iJ2Pp2GKxuPxeGJpPJlK2WuTezqjOZJjg5ksNnsDiCMHgCG5Xi+H1C4UiGtFqZ5LkesE6aSEMBVSSI+G7YntAPl85+dv+gOnejTA+9WqSI+NMHHDEn8oeh9Nob54atFpjAPLbr0lcOSt3T6aB76x9PhtHAxDFcI2vGZ71WR8ky1VMoS/Mc5BPLxnQTLNmhzZo8xxPE4EJEk4DJVgKSpCs6krOoAG0rHrdE7AASQEIQ3EojFaPogB5fCKSgGhfB4GimCESw1liABdOpnFYCQIHgPAUnPYMXk7W8bzA3kwJQ3Q0N0LAAHYlJmPAAGpVMXVYaH6OAAQxFVQV2IA==)

We should also cache `pinNumber.ToString()` in a local since it's common between the branches.

Reply:
ok @ahsonkhan will update, thanks